js/Chart.js

```diff
@@ -1,485 +1,545 @@
-      d3.select(_this.data.getDatum(j, i).columnEl).attr('class', columnClass)
+    if (end >= this.xEndEdge) {
+      this.$xEnd.addClass('hidden')
+    } else {
+      this.$xEnd.removeClass('hidden')
     }
-  }
 
-  this.updateSelectionText()
-}
+    // move selected dots
+    var _this = this
+    this.layers.each(function (seriesIndex, i) {
+      var seriesExtent = _this.data.getSeriesExtent(i)
+      if (_this.selectedX >= seriesExtent[0] && _this.selectedX <= seriesExtent[1]) {
+        d3.select(this).selectAll('.selected-dot')
+          .attr('cx', _this.xPositionLine(_this.data.getDatum(i, _this.selectedX)))
+          .attr('cy', _this.yPosition(_this.data.getDatum(i, _this.selectedX)))
+          .style({'opacity': '1'})
+      } else {
+        d3.select(this).selectAll('.selected-dot').style({'opacity': '0'})
+      }
+    })
+
+    // add selected class to the appropriate columns
+    var columns = this.data.getIndexCount()
+    for (var i = 0; i < columns; i++) {
+      var columnClass = 'column'
+      if (_this.selectedX === i) {
+        columnClass = 'column selected'
+      }
 
-Chart.prototype.updateSelectionText = function() {
-  // a focusedSeriesIndex >= to the ySeries length means use the total
-  var showTotal = this.focusedSeriesIndex >= this.data.getSeriesCount()
-  var chartYSeries = showTotal ? this.data.getSeriesCount() - 1 : this.focusedSeriesIndex
-  var thisPoint = this.data.getDatum(chartYSeries, this.selectedX)
-
-  var thisYLabel = ''
-  var thisYColor = showTotal ? this.colorBase : this.getSeriesColor(this.data.getSeries(thisPoint.ySeries).seriesIndex)
-  if (this.data.getSeriesCount() > 1) {
-    var pageYSeries = this.getChartSeries()[chartYSeries]
-    thisYLabel = showTotal ? 'total' : this.pageController.getSeriesName(pageYSeries)
+      for (var j = 0; j < _this.data.getSeriesCount(); j++) {
+        let el = _this.data.getDatum(j, i).columnEl
+        if (el) {
+          d3.select(el).attr('class', columnClass)
+        }
+      }
+    }
+
+    this.updateSelectionText()
   }
 
-  var seriesExtent = this.data.getStackedExtentForIndex(this.selectedX)
-  var seriesTotal = seriesExtent[1] + seriesExtent[0]
-  var thisValue = showTotal ? seriesTotal : thisPoint.yRaw
-  var thisValueFormatted = this.params.rounding === 'on' ? Utils.getRoundedValue(thisValue, this.yRange) : thisValue
+  updateSelectionText(): void {
+    // a focusedSeriesIndex >= to the ySeries length means use the total
+    var showTotal = this.focusedSeriesIndex >= this.data.getSeriesCount()
+    var chartYSeries = showTotal ? this.data.getSeriesCount() - 1 : this.focusedSeriesIndex
+    var thisPoint = this.data.getDatum(chartYSeries, this.selectedX)
 
-  // update selection
-  this.$selectionYLabel.text(thisYLabel).css('color', thisYColor)
-  this.$selectionXLabel.text(thisPoint.xLabel)
-  this.$selectionValue.text(thisValueFormatted).css('color', thisYColor)
-}
+    var thisYLabel = ''
+    var thisYColor = null
```

Comment:
There's currently an issue in Master where hovering above the stacked bars (not on a specific bar) — so it shows "Total" — doesn't switch the color of "Total" back to black. (Or, back to white when it's a dark background). Since we're refactoring here, we should make sure this issue is fixed.

I haven't been able to test this branch locally yet, but it looks like you already fixed it with `null`? (For both light/dark backgrounds?) If not, this should also work:

```
var thisYColor = this.pageController.getPageColor() === 'dark' ? this.colorDark : this.colorLight
```